main.ts

```diff
@@ -1,129 +1,197 @@
-        break;
-    }
-});
+//action / phase listeners:
+thousand.events.addListener('action', onPlayerAction);
+thousand.events.addListener('phaseUpdated', onPhaseUpdated);
 
 thousand.init();
 
-// _.chain([
-//     () => thousand.registerPlayer('adam'),
-//     () => thousand.registerPlayer('alan'),
-//     () => thousand.registerPlayer('pic'),
-//     () => thousand.bid('alan', 110),
-//     () => thousand.pass('pic'),
-//     () => thousand.pass('adam'),
-//     //alan is winner
-//     () => thousand.shareStock(getCardsByPlayer('alan')[0], 'adam'),
-//     () => thousand.shareStock(getCardsByPlayer('alan')[0], 'pic'),
-//     //battle:
-//     () => thousand.throwCard(getCardsByPlayer('alan')[0], 'alan'),
-// ]).map(action => action()).value();
+_.chain([
+    () => thousand.registerPlayer('adam'),
+    () => thousand.registerPlayer('alan'),
+    () => thousand.registerPlayer('pic'),
+    () => thousand.bid('alan', 110),
+    () => thousand.pass('pic'),
+    () => thousand.pass('adam'),
+    // //alan is winner
+    () => thousand.shareStock(getCardsByPlayer('alan'), 'adam'),
+    () => thousand.shareStock(getCardsByPlayer('alan'), 'pic'),
+    // //battle:
+    () => thousand.throwCard(getCardsByPlayer('alan'), 'alan'),
+    () => thousand.throwCard(getCardsByPlayer('pic'), 'pic'),
+    () => thousand.throwCard(getCardsByPlayer('adam'), 'adam'),
+    () => thousand.throwCard(getCardsByPlayer('adam'), 'adam')
+]).map(action => action()).value();
+
+console.log(thousand.getState());
+
+function getCardsByPlayer(player) {
+    return thousand.getState().cards[player][0];
+}
+
+//------------------------------------------------------------------------
+
+function onPlayerAction(action: Action) {
+    const actionHandlers = {
+        [THROW_CARD]: () => {
+            const throwAction = action as ThrowCard;
+            console.log(`ACTION: ${throwAction.player} has thrown ${toString(throwAction.card)}`);
+        },
+        [BID]: () => {
+            const bidAction = action as Bid;
+            if (bidAction.pass) {
+                console.log(`ACTION: ${bidAction.player} has passed`);
+            } else {
+                console.log(`ACTION: ${bidAction.player} has bidded ${bidAction.bid}`);
+            }
+        },
+        [REGISTER_PLAYER]: () => {
+            const registerAction = action as RegisterPlayer;
+            console.log(`ACTION: ${registerAction.id} joined the table`);
+        },
+        [SHARE_STOCK]: () => {
+            const shareStock = action as ShareStock;
+            console.log(`ACTION: sharing ${toString(shareStock.card)} with opponent: ${shareStock.player}`);
+        }
+    };
+    actionHandlers[action.type] && actionHandlers[action.type]();
+}
+
+function onPhaseUpdated(next: Function, phaseInit: boolean) {
+    const state: Game = thousand.getState();
+
+    const phaseHandlers = {
+        [Phase.REGISTERING_PLAYERS_START]: () => {
+            console.log('show registering player view');
```

Comment:
Maybe we could use some abstraction for logging? So you easily change the log level or disable it completely.